### PR TITLE
BINDING 3.4.4.1 - DEFLATE Encoding - DSA signature algorithm

### PR DESCRIPTION
**Pull request by brjeter**
#### What does this PR do (if it's not clear from the Title)? Please include any specification references that apply.
- Changes MetadataConfigurationParser to take an `EntitiesDescriptor` that contains multiple `EntityDescriptor` elements.
- Changes SimpleSign to optionally take in an `entityId` which it will use to look up a particular key in the keystore for its signing.
- Changes `TestCommon` to allow for multiple SPs. **Note: This will be reworked in a second commit shortly using a kotlintest interceptor or spec configuration to configure the current SP. The current code is a little hacky.**
- Added a Redirect test to verify the usage of a DSA signing algorithm.

#### Checklist:
- [ ] SAML Spec Table of Contents documentation updated
- [X] Unit Tests Added/Modified

### CODE REVIEW

File: ctk/common/src/main/kotlin/org/codice/compliance/utils/TestCommon.kt [hunk 1]
```diff
@@ -91,19 +91,20 @@ class TestCommon {
...
-        val SP_ENTITY_INFO = checkNotNull(spMetadata[SP_ISSUER])
+        const val DEFAULT_SP_ISSUER = "https://samlhost:8993/services/saml"
+        const val DSA_SP_ISSUER = "https://samlhostdsa:8993/services/samldsa"
+
+        private val spMetadata = Common.parseSpMetadata()
+        val DEFAULT_SP_ENTITY_INFO = checkNotNull(spMetadata[DEFAULT_SP_ISSUER])
+        val DSA_SP_ENTITY_INFO = checkNotNull(spMetadata[DSA_SP_ISSUER])
+
+        var CURRENT_SP_ISSUER = DEFAULT_SP_ISSUER
+        var CURRENT_SP_ENTITY_INFO = DEFAULT_SP_ENTITY_INFO
```
Comment (brjeter) [inline ctk/common/src/main/kotlin/org/codice/compliance/utils/TestCommon.kt]: This part will be reworked in the second commit. See the PR description for details.

File: ctk/common/src/main/kotlin/org/codice/compliance/utils/XMLDecryptor.kt [hunk 2]
```diff
@@ -37,7 +37,7 @@ class XMLDecryptor {
             org.apache.xml.security.Init.init()
         }
 
-        private const val ENCRYPTION_FILE_NAME = "encryption.properties"
+        private const val ENCRYPTION_FILE_NAME = "samlhost-encryption.properties"
```
Comment (brjeter) [inline ctk/common/src/main/kotlin/org/codice/compliance/utils/XMLDecryptor.kt]: Going to fix this to not be hard-coded. I must've missed this.

File: ctk/idp/src/main/kotlin/org/codice/compliance/web/sso/RedirectSSOTest.kt [hunk 3]
```diff
@@ -130,14 +135,49 @@ class RedirectSSOTest : StringSpec() {
             }
         }
 
+        "Redirect AuthnRequest Using DSA1 Signature Algorithm" {
```
Comment (brjeter) [inline ctk/idp/src/main/kotlin/org/codice/compliance/web/sso/RedirectSSOTest.kt]: I will update the test with description and spec reference.

File: ctk/idp/src/main/kotlin/org/codice/compliance/web/sso/RedirectSSOTest.kt [hunk 4]
```diff
@@ -130,14 +135,49 @@ class RedirectSSOTest : StringSpec() {
...
+            Log.debugWithSupplier { "Redirect AuthnRequest Using DSA1 Signature Algorithm" }
+            CURRENT_SP_ISSUER = DSA_SP_ISSUER
+            CURRENT_SP_ENTITY_INFO = DSA_SP_ENTITY_INFO
```
Comment (brjeter) [inline ctk/idp/src/main/kotlin/org/codice/compliance/web/sso/RedirectSSOTest.kt]: This part will not be needed in the second commit.

File: library/src/main/java/org/codice/security/saml/MetadataConfigurationParser.java [hunk 5]
```diff
@@ -30,6 +31,7 @@
 public class MetadataConfigurationParser {
```
Comment (brjeter) [inline library/src/main/java/org/codice/security/saml/MetadataConfigurationParser.java]: I will probably look into donating the changes to `MetadataConfigurationParser` back to DDF.

File: library/src/main/java/org/codice/security/sign/SystemCrypto.java [hunk 6]
```diff
@@ -42,16 +42,20 @@
...
   private final String encryptionAlias;
 
-  public SystemCrypto() throws IOException {
+  public SystemCrypto(String hostname) throws IOException {
     this.passwordEncryption = null;
     // new EncryptionServiceImpl()
 
-    Properties sigProperties = createProperty("encryption.properties");
+    if (hostname == null) {
+      hostname = "samlhost";
```
Comment (brjeter) [inline library/src/main/java/org/codice/security/sign/SystemCrypto.java]: I will make this a constant somewhere. Something like `DEFAULT_HOSTNAME` or something.

File: library/src/test/java/org/codice/security/sign/SimpleSignTest.java [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
...YgRGVtbyBSb290IENBMTEwLwYJKoZIhvcNAQkBFiJlbWFpbEFkZHJlc3M9ZGRmcm9vdGNhQGV4YW1wbGUub3JnMB4XDTE4MDUwOTE1NDUyOFoXDTE5MDUwOTE1NDUyOFowTDELMAkGA1UEBhMCVVMxCzAJBgNVBAgMAkFaMQwwCgYDVQQKDANEREYxDDAKBgNVBAsMA0RldjEUMBIGA1UEAwwLc2FtbGhvc3Rkc2EwggG4MIIBLAYHKoZIzjgEATCCAR8CgYEAvSzXEhq+o1re6lKMzgYepZ2+18xiZChKAyhfSIzpAUXR8kmYKIEz/tIW+rXv+t7CSGoiKgIj8lHoEwLQ4wnaOg9u51PE5xV9f0ig9vCQGzwXq8yi3YBKfWDCSWXN0cEsBP6Xfvf3boFbWjzMvwuE2VfkAmQzOhL3372mGtuMBZ8CFQDa0xC8ejL6JCRpt1JVU2DVY9/fdwKBgQCgMcRXMpRlw01I4V2DbaynZPID1H2FdPOnPFqhVgXK13fJe0i0oa3hNZII7UrdCMI7MZ3eMShXbKDWdb0RnAPXleVdnWsTWZU4puozmuBI0AH7BUfN7tQPFVmwqPXpL4DJoKH2p3dcsUnVAKHa+SehGnVA0ZiHPnK7sdrb6Ns52gOBhQACgYEApH128rhv7r3q8G1eCSYRS8bEO2Rel9FV74igdcfpvD5FqjOhQ1K5zxKVnqQbqqaNTuRlhWb9C94LJB/KozNXWf8TCP/gbz6pd0nJGLP84lpke/ffblbCS05XtNtg4X1qOu+HdNO9xedEoM0K0QlyF09LCZNunr4CTj5z8uPDICCjgYEwfzAJBgNVHRMEAjAAMCcGCWCGSAGG+EIBDQQaFhhGT1IgVEVTVElORyBQVVJQT1NFIE9OTFkwHQYDVR0OBBYEFJgSNpRmsXQSHbIXORCCPipDHKFGMB8GA1UdIwQYMBaAFOFUx5ffCsK/qV94XjsLK+RIF73GMAkGA1UdEQQCMAAwDQYJKoZIhvcNAQEFBQADgYEAQW4bQOzumWE3erDYDy11O3wbk2zL2TN72DdDn5p58QTA3cDhlOBCdyeCBtRJ0RbSDWEswjwMiXuypYzvNLZ4z7QqlsvoCfE7SI6jBNkAhqP04EuEjPBx8NZUQsYp4tTeJDPGHcY7xdl1SqfGcQ1ofSGGl/EzoSmHQxH01yhpYhU=";
```
Comment (brjeter) [inline library/src/test/java/org/codice/security/sign/SimpleSignTest.java]: Unfortunately there could be maintenance cost with these hardcoded cert strings since we'll have to change them whenever we change the certs in the keystore. I started down the path of doing it dynamically, and decided that was a little much for a unit test.

File: ctk/idp/src/main/kotlin/org/codice/compliance/web/sso/RedirectSSOTest.kt (hunk 3 above)
Comment (blen-desta) [reply ctk/idp/src/main/kotlin/org/codice/compliance/web/sso/RedirectSSOTest.kt]: For the negative tests I did `Doc section: test name` like `Core 3.1.4: POST incorrect x tests` and same for the log message below

File: ctk/idp/src/main/kotlin/org/codice/compliance/web/sso/RedirectSSOTest.kt [hunk 8]
```diff
@@ -130,14 +135,49 @@ class RedirectSSOTest : StringSpec() {
...
+            CoreAuthnRequestProtocolVerifier(authnRequest, samlResponseDom).apply {
+                verify()
+                verifyAssertionConsumerService(finalHttpResponse)
+            }
+            SingleSignOnProfileVerifier(authnRequest, samlResponseDom).apply {
+                verify()
+                verifyBinding(finalHttpResponse)
+            }
+
+            CURRENT_SP_ISSUER = DEFAULT_SP_ISSUER
```
Comment (blen-desta) [inline ctk/idp/src/main/kotlin/org/codice/compliance/web/sso/RedirectSSOTest.kt]: ❓ Do we really need to set it back? If this test passes that mean the IdP supports DSA so we can use that for the remaining tests.

File: library/src/main/java/org/codice/security/saml/MetadataConfigurationParser.java (hunk 5 above)
Comment (blen-desta) [reply library/src/main/java/org/codice/security/saml/MetadataConfigurationParser.java]: You should 💯

File: ctk/common/src/main/kotlin/org/codice/compliance/utils/TestCommon.kt [hunk 9]
```diff
@@ -222,3 +270,19 @@ class TestCommon {
         }
     }
 }
+
+class LazyVar<T>(val init: () -> T) : ReadWriteProperty<Any?, T> {
```
Comment (bakejeyner) [inline ctk/common/src/main/kotlin/org/codice/compliance/utils/TestCommon.kt]: Cool

File: ctk/idp/src/main/kotlin/org/codice/compliance/web/sso/RedirectSSOTest.kt [hunk 10]
```diff
@@ -135,15 +131,14 @@ class RedirectSSOTest : StringSpec() {
             }
         }
 
-        "Redirect AuthnRequest Using DSA1 Signature Algorithm" {
+        "Bindings 3.4.4.1: Redirect AuthnRequest Using DSA1 Signature Algorithm".config(
+                extensions = listOf(UseDSASigningSP)) {
```
Comment (bakejeyner) [inline ctk/idp/src/main/kotlin/org/codice/compliance/web/sso/RedirectSSOTest.kt]: Cool

File: ctk/common/src/main/kotlin/org/codice/compliance/utils/TestCommon.kt [hunk 11]
```diff
@@ -90,20 +99,60 @@ class TestCommon {
         const val PRIVATE_KEY_PASSWORD =
                 "org.apache.ws.security.crypto.merlin.keystore.private.password"
 
-        private val DEPLOY_CL = getDeployDirClassloader()
-        private val spMetadata = Common.parseSpMetadata()
+        private const val DEFAULT_SP_ISSUER = "https://samlhost:8993/services/saml"
+        private const val DSA_SP_ISSUER = "https://samlhostdsa:8993/services/samldsa"
+        @JvmField
```
Comment (blen-desta) [inline ctk/common/src/main/kotlin/org/codice/compliance/utils/TestCommon.kt]: ❓ Just curious why we don't want getters and setters for `currentSPIssuer`?

File: ctk/common/src/main/kotlin/org/codice/compliance/utils/TestCommon.kt [hunk 12]
```diff
@@ -90,20 +99,60 @@ class TestCommon {
...
-            SamlProtocol.Binding.values()
-                    .associate {
-                        it to spInfo.getAssertionConsumerService(it)?.url
-                    }
+                currentSPIssuer = DEFAULT_SP_ISSUER
+                currentSPEntityInfo = DEFAULT_SP_ENTITY_INFO
+            }
+        }
+
+        @JvmStatic
```
Comment (blen-desta) [inline ctk/common/src/main/kotlin/org/codice/compliance/utils/TestCommon.kt]: ❓ Reading a bit about this. 
> Specifies that an additional static method needs to be generated from this element if it's a function.

Why is this necessary? Doesn't the fact that it's in a companion object make it static?

**Comment by brjeter** [comment]
@bakejeyner, I looked into using the custom setter. That would've worked except that "property1" in this case is `currentSPIssuer` which has to be accessible by Java. And you can't have custom setters with a `@JvmField` 😞

**Comment by brjeter** [comment]
@blen-desta That's fair 😉

File: library/src/test/resources/ddf-sp-metadata.xml [hunk 13]
```diff
@@ -1,28 +1,30 @@
...
-    <md:SingleLogoutService Binding="urn:oasis:names:tc:SAML:2.0:bindings:HTTP-Redirect" Location="https://localhost:8993/services/saml/logout"/>
-    <md:SingleLogoutService Binding="urn:oasis:names:tc:SAML:2.0:bindings:HTTP-POST" Location="https://localhost:8993/services/saml/logout"/>
-    <md:SingleLogoutService Binding="urn:oasis:names:tc:SAML:2.0:bindings:SOAP" Location="https://localhost:8993/services/saml/logout"/>
-    <md:AssertionConsumerService Binding="urn:oasis:names:tc:SAML:2.0:bindings:HTTP-Redirect" Location="https://localhost:8993/services/saml/sso" index="0"/>
-    <md:AssertionConsumerService Binding="urn:oasis:names:tc:SAML:2.0:bindings:HTTP-POST" Location="https://localhost:8993/services/saml/sso" index="1"/>
-    <md:AssertionConsumerService Binding="urn:oasis:names:tc:SAML:2.0:bindings:PAOS" Location="https://localhost:8993/services/saml/sso" index="2"/>
-  </md:SPSSODescriptor>
-</md:EntityDescriptor>
\ No newline at end of file
+<md:EntitiesDescriptor xmlns:md="urn:oasis:names:tc:SAML:2.0:metadata">
```
[the remark this answers is not in the document]
Comment (brjeter) [reply library/src/test/resources/ddf-sp-metadata.xml]: This is used for the unit tests only. I will modify the name of the xml file to make that more explicit.

File: deployment/distribution/src/main/resources/samlconf-sp-metadata.xml [hunk 14]
```diff
@@ -1,28 +1,58 @@
...
+          </ds:X509Data>
+        </ds:KeyInfo>
+      </md:KeyDescriptor>
+      <md:KeyDescriptor use="encryption">
+        <ds:KeyInfo xmlns:ds="http://www.w3.org/2000/09/xmldsig#">
+          <ds:X509Data>
+            <ds:X509Certificate>
+              MIICsDCCAhmgAwIBAgIGAWH8X/0MMA0GCSqGSIb3DQEBCwUAMIGEMQswCQYDVQQGEwJVUzELMAkGA1UECBMCQVoxDDAKBgNVBAoTA0RERjEMMAoGA1UECxMDRGV2MRkwFwYDVQQDExBEREYgRGVtbyBSb290IENBMTEwLwYJKoZIhvcNAQkBFiJlbWFpbEFkZHJlc3M9ZGRmcm9vdGNhQGV4YW1wbGUub3JnMCAXDTE4MDMwNTE3MzMyNFoYDzIxMTgwMzA1MTczMzI0WjATMREwDwYDVQQDDAhzYW1saG9zdDCCASIwDQYJKoZIhvcNAQEBBQADggEPADCCAQoCggEBAKvxMUgWYH/1BBbS6gAV8zmMgqrhlnNvbn8nFyHewcNC4lkYmh3TUXbxxymfvCduOSDKZh3DJN65oL1PKc8gozZOR978VaxizPCjJNUTCzXQ4pfyHE0GIOkynZT63ZoJhlbJaD7sosbcXdD18CZZ/c3lS+4W2XeMClQoHNgmho14i9I/tf2DJ5j/dIrOk5UJ4mg9xTA3TFnLPLpPiMtvW/QQmfK1z59LBIFnGr9OchiWyJPxe8ND0LEsDv0NQ0CP+LhdM3wNuZnxjV5IdPXp1GKF24ouBUa3kGtYrdu+nj2icpp/tVjOGFnBuswH7rckKEzcmYAdW1buG00AmydLrg8CAwEAAaMXMBUwEwYDVR0RBAwwCoIIc2FtbGhvc3QwDQYJKoZIhvcNAQELBQADgYEAoPIq6VIT/pR0VF2bejTog3JEM4y01qI30CXjP58PVHjPJcQxDSQCQc/GFPWfFNaSwg9LR0nPXbhRPnwpveROEzYXabV60e+8Ny9VsGu7Siu7cdcuM5ZpTdSamtioHZ1LDw4lvtLgXTiYmipm55SY9DIhxVgULylq1tOSblgS66M=
```
[the remark this answers is not in the document]
Comment (brjeter) [reply deployment/distribution/src/main/resources/samlconf-sp-metadata.xml]: Tested this on Keycloak because, unlike DDF, it actually encrypts things using this cert. It failed like this AND when I copied the DSA signing cert in as the encrypting cert 😞 Keycloak doesn't support DSA algorithm for encrypting. Idk if that's common, but the SAML Spec doesn't say anything about it so Keycloak is fine to do that. In order to get it to pass on Keycloak, I copied the default `samlhost` cert in as the encrypting cert for `samlhostdsa` with a comment that we can investigate more when we change DDF to support SAML encryption.

**Comment by brjeter** [comment]
build now
